torch/csrc/autograd/engine.h

```diff
@@ -68,6 +72,29 @@ struct TORCH_API Engine {
   std::vector<std::function<void()>> final_callbacks_;
   // To protect reads and writes to final_callbacks_
   std::mutex post_callbacks_lock_;
+
+  struct ThreadPoolShared {
```

Comment:
why do we need to add a thread pool implementation when we already have thread pools in PyTorch?

Reply:
`c10/core/thread_pool.h` model is that tasks are thread safe and can be run on any thread which is available. The autograd model is different: every task must be associated with a specific thread for its device because they are not thread safe to run concurrently with other tasks on the same device.

Furthermore, the parallelism that is being taken advantage of here is purely to ensure we can enqueue CUDA kernels quickly enough; no real kernel-based parallelism takes place (we need more threads because of the details of #6959; sometimes threads get themselves into blocked states and we need to spin up another worker thread to deal with them.)